### DISCUSSION TITLE
Source generator that doesn't run interactively in editor

### DISCUSSION THREAD

**Discussion by Aidiakapi** [dotnet/roslyn#53423] [Q&A]
Both the [design document](https://github.com/dotnet/roslyn/blob/main/docs/features/source-generators.md) and the [cookbook](https://github.com/dotnet/roslyn/blob/main/docs/features/source-generators.cookbook.md) seem to suggest that the default behavior is that source generators only affect the build, and have no IDE integration.

We have a fairly large project (~10 second csc.exe for this specific project in the solution), which is currently doubled, because we essentially compile the project twice. Once in a pre-build step which invokes the Roslyn APIs to perform a compilation, does a bunch of analysis on the types, and emits a single source file of about 15MB in size, after which it will perform the actual build.

This sounds like a trivial case for source generation, it's fully additive, it doesn't need any intellisense support, and solely depends on the source syntax trees. The problem is that it's way too slow to run in the editor (~7 seconds), and while I suspect a 2x improvement there is possible, it's unlikely to go down to anything < 1 second.

Is there a way to run the source generator such that it doesn't run interactively at all, and only in specific cases, such as when debugging? I'm currently hacking around it in the `.csproj`, so that it only runs at build time:

```xml
  <Target Name="Reference source generator" BeforeTargets="BeforeBuild">
    <ItemGroup>
      <Analyzer Include="..\path\to\source_generator.dll" />
    </ItemGroup>
  </Target>
```

**Comment by CyrusNajmabadi** [comment]
> Both the design document and the cookbook seem to suggest that the default behavior is that source generators only affect the build, and have no IDE integration.

Any documentation to that effect is unfortunately wrong :(   Sorry about taht.  Source generators do run in the IDE and are necessary for correctness and to ensure users don't get things like tons of error squiggles everywhere because they ahve code that references the generated code.

**Reply by Aidiakapi** [reply]
I know why they should run in the IDE, I just thought I had to implement some of the syntax notifier stuff to support it.

Regardless, is there an up-to-date documentation somewhere? It seems like the only documentation is those two things I linked, and the two blog posts.

**Comment by jasonmalinowski** [comment]
And we are working on APIs to better allow for better control of this, but those aren't available yet. (Tagging @chsienki for awareness.)

Can you describe a bit more what your generator is doing? We're generally not wanting people to implement caching, but in certain limited cases it may help until we get those APIs available, which may be helpful here.

**Reply by Aidiakapi** [reply]
I responded with that below btw.

Speaking of the caching APIs, I think for a lot of cases where source generators will be useful, it's going to be about taking the codebase, analysing parts of it, building a model, and then looking at specific events that might change this model, such as a new type being added, an attribute applied, a member added, etc.

While we have several candidates for source generators, I'm weary of using them due to the performance implications. We have this one that runs during builds only as described below, that now runs in about 900ms. I've also added one that runs in the IDE, that generates various boilerplate-y things for a log category enum. In this case, it only has to run if the enum has members modified.

**Reply by CyrusNajmabadi** [reply]
> In this case, it only has to run if the enum has members modified.

As an example of the complexity here.  What does it mean to say "has member modified"? 

Does it mean: they're syntactically different?  Does that include if they're at different whitespace positions?  
Does it mean: they're semantically different?  What is a semantic difference?  Would it include things like if the value of the enum member was a computed constant, and that computed constnat changed value due to changes elsewhere?
etc. etc. etc.

**Reply by Aidiakapi** [reply]
I understand that a fully accurate query is ambiguous and will likely be slow. The way I see it is like a broad phase in a physics simulation. The goal is not to get the precise interactions in a region, but to very rapidly eliminate the majority of potential interactions.

While I don't have the answers ofc, I think that most people will be working on the semantic model, occasionally dipping into the syntactic model for things like trivia. If you could register callbacks for any change (without knowing the change) in a specific namespace, specific type, or specific method, that might already be enough for a lot of cases.

I appreciate the desire to make a great API that handles all cases, and does a lot of the heavy lifting.

Personally, for the source generators that run in-IDE, they both respond to specific changes in a specific file. So only regenerating when that file has changed is sufficient. Is there a way to do that at the moment?

**Reply by CyrusNajmabadi** [reply]
>Personally, for the source generators that run in-IDE, they both respond to specific changes in a specific file. So only regenerating when that file has changed is sufficient. Is there a way to do that at the moment?

Yup!  That is possible today (And Roslyn uses this ourselves).  See what we do in CachingSourceGenerator, particularly here: https://github.com/dotnet/roslyn/blob/47340e077cf9fbd6c51daaca7e07956393691b30/src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/CachingSourceGenerator.cs#L46-L57

The general idea is that files have a checksum based on their content.  So if you want to return the same value as long as the checksum hasn't changed you can.  If you return the *exact* same SourceText instance (not just the same contents, but the same object itself) we can also optimize that to then do no additional work for things like parsing that file, etc.

**Reply by jasonmalinowski** [reply]
And for anybody seeing @CyrusNajmabadi's comment now, we've now created a new, more advanced incremental generator API that doesn't mean you need to use something like CachingSourceGenerator now.

**Comment by Aidiakapi** [comment]
It's essentially generating serialization code for a networked game. It iterates over most types, and their fields recursively, and then generates specific serializers for them, which have to work on AoT compiled platforms.

Ideally we'd avoid the issue altogether, by operating as a post-build task, where we can generate the serialized as a separate DLL. However, there are various obstacles in the way or that at the moment. So as an alternative, we thought that since we already use the Roslyn APIs, we might be able to relatively inexpensively switch it to use source generators. The problem is of course the performance.

We could technically seen try to do an incremental style of generation, but that'd add way too much complexity to the system for no real benefit.

I'm open for any suggestions though, and thanks already for the responses :)

**Comment by sake402** [comment]
Would really love an answer to this too. Have a source generator that generates a bunch of codes, up to 1000 classes currently and this seems to slow VS2022 down a lot while editting. Tracing RoslynCodeAnalysis reveal spending so much time calling our custom generators and Microsoft.Extensions.Logging.Generators.LoggerMessageGenerator, RazorSourceGenerator, JsonGenerator over and over and makes up for a bad editting experience in VS2022.

![image](https://user-images.githubusercontent.com/20999608/164685811-3e002b95-accc-47c3-8241-912257ef5976.png)

Would propose a way to flag a generator as only running during build, if the user doesn't need intellisense from code generated by it.

Something along the line of
```
    <ProjectReference Include="MyGenerator.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="false" RunInEditor="false" />
```

**Reply by CyrusNajmabadi** [reply]
@sake402 can you make this an incremental generator?

**Reply by sake402** [reply]
@CyrusNajmabadi It is an IncrementalGenerator

**Reply by Aidiakapi** [reply]
We're still using the trick in OP. Only register the source generator in a pre-build step in the csproj.

**Reply by sake402** [reply]
@Aidiakapi  We are hacking around like this

```C#
        var syntaxes = context.SyntaxProvider.CreateSyntaxProvider(DoFilter, (syntax, cancellation) => syntax);

        protected bool DoFilter(SyntaxNode node, CancellationToken token)
        {
            var processName = System.Diagnostics.Process.GetCurrentProcess().MainModule.ModuleName;
            if (processName.Contains("VBCSCompiler")|| processName.Contains("dotnet"))
            {
                //perform your default filtering
            }
            return false; //don't run generator if not VBCSCompiler and not dotnet
        }

```

Your method is probably better as it doesnt mess with the generator itself

**Reply by CyrusNajmabadi** [reply]
@sake402 what is the perf overhead coming from then in that case?  Is your code running unnecessarily?  If so, we should just fix that bug rather than trying to have a system that has entirely different semantics while editing, versus while compiling.

**Reply by CyrusNajmabadi** [reply]
@Aidiakapi that won't work for things like if csc decidedes to not use the compiler server, or if we ever run things like SGs in a sandbox.  Can you clarify the above questions around what is going on even if you have incremental generators?  WHere is the time being spent?

**Reply by sake402** [reply]
> Is your code running unnecessarily?

I would assume so, since everytime I capture a trace on RoslynCodeAnalysis service, the stack is always filled with Source generator codes.

After using the DoFilter method above, things get quite better, my source generator no longer runs in RoslynCodeAnalysisService, but the service is still having significant CPU usage. Here is the stack.

![image](https://user-images.githubusercontent.com/20999608/164754389-fa92769f-1104-4606-82c6-3bd7cee45a27.png)

**Reply by jasonmalinowski** [reply]
@sake402: which version of Visual Studio was this with? That may be the issue we tried to address with https://github.com/dotnet/roslyn/pull/60531.
